### ISSUE TITLE
Error when querying for multiple invoices

### ISSUE THREAD

**Issue by IamDavidovich** [sportngin/active_zuora#38]
I have an error popping up any time I request more than one invoice:

```
> ActiveZuora::Account.invoices.all
Savon::SOAP::Fault: (fns:INVALID_VALUE) Can only query one invoice body at a time.: select Id, AccountId, AdjustmentAmount, Amount, AmountWithoutTax, Balance, Body, Comments, CreatedById, CreatedDate, CreditBalanceAdjustmentAmount, DueDate, IncludesOneTime, IncludesRecurring, IncludesUsage, InvoiceDate, InvoiceNumber, LastEmailSentDate, PaymentAmount, PostedBy, PostedDate, RefundAmount, Status, TargetDate, TaxAmount, TaxExemptAmount, TransferredToAccounting, UpdatedById, UpdatedDate from Invoice where AccountId = 'account_id_was_here'
```

It's fine if I limit it in a way that I'm only requesting a single invoice:

```
> ActiveZuora::Invoice.where(invoice_number: 'INV00000001').first
=> #<ActiveZuora::Invoice:0x007fbbee188088
 @account_id="account_id_was_here",
  ...
```

Is there a way of fixing this without having to manually implement the Invoice model?

**Comment by anfleene** [comment]
Yeah thats a limitation of the Zuroa api iirc. They won't send you the body unless you just ask for a single invoice. It would require basically requerying  for the body after we've created the AZ invoice object so yeah I think we'd have to change the invoice model. We already have a [hook for customizing](https://github.com/sportngin/active_zuora/blob/master/lib/active_zuora/generator.rb#L144) it though. So patching it to work this way is possible.

**Comment by IamDavidovich** [comment]
Cool. Thanks for the info! I'll have a look and push up a PR if I get something working.

**Comment by IamDavidovich** [comment]
So, I've got [this far](https://github.com/IamDavidovich/active_zuora/commit/bee5e0e80f2df5c827e54a6ad6b75e489ffa64b5). It's loading the body as a separate call when it's accessed, so it solves the issue when querying for multiple invoices. 

Unfortunately, every field is accessed as part of the object initialization, which means that it's making a request for every invoice in a collection and basically just stalls if you request more than a few invoices at a time. 

I'm tempted to just have the body accessed by a different method (eg. `pdf`) but would prefer to be able to lazy-load `body`. I don't suppose you can point me towards any way of skipping `body` at initialization time?

**Comment by anfleene** [comment]
yeah sure you would just need to add it as an [exlucsion to the genorator](https://github.com/sportngin/active_zuora/blob/master/lib/active_zuora/generator.rb#L144-L147)

```ruby
      customize 'Invoice' do
        include Generate
        exclude_from_queries :regenerate_invoice_pdf, :body
      end
```

**Comment by IamDavidovich** [comment]
Yep, I've got the exclusion in that commit I linked to above. That works great and allows you to fetch multiple invoices, but it means that unless you write a specific query to fetch it the `body` field is never populated. 

What I'm trying to do is lazy-load the body field. So, when the invoices are queried the body field is not loaded, but when it's actually accessed a separate request is made to fetch the body. The issue I have with the current implementation is that the first thing that happens when an object is loaded is that every field is accessed, so the lazy loading is triggered.

I've tried a few different things, but without much luck, so what I might do for now is block it from the queries and then add in a `pdf` method to access that data if needed.

**Comment by anfleene** [comment]
Ah sorry I missunderstood. I'll look a bit closer at your implementation later today and see if any solution pops out at me.

**Comment by IamDavidovich** [comment]
Thanks! I've put up #41 in the meantime.

**Comment by anfleene** [comment]
I just released 2.2.1 Which should solve this problem. Thanks for reporting it.

**Comment by IamDavidovich** [comment]
No problem. Thanks for fixing it!
